### PR TITLE
SDK - Refactoring - Replaced the TypeMeta class

### PR DESCRIPTION
**Pull request by Ark-kun**
The PipelineParam no longer exposes the private TypeMeta class
Fixes https://github.com/kubeflow/pipelines/issues/1420

The refactoring PR is part of a series of PR which unifies the metadata and specification types.

All the tests continue to pass.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1930)

### CODE REVIEW

File: sdk/python/kfp/dsl/_pipeline_param.py [hunk 1]
```diff
@@ -136,7 +135,7 @@ class PipelineParam(object):
   value passed between components.
   """
   
-  def __init__(self, name: str, op_name: str=None, value: str=None, param_type: TypeMeta=TypeMeta(), pattern: str=None):
+  def __init__(self, name: str, op_name: str=None, value: str=None, param_type = None, pattern: str=None):
```
Comment (ajchili) [inline sdk/python/kfp/dsl/_pipeline_param.py]: Should `param_type` have a type?

**Comment by numerology** [comment]
> The PipelineParam no longer exposes the private TypeMeta class
> Fixes #898
> 
> The refactoring PR is part of a series of PR which unifies the metadata and specification types.
> 
> All the tests continue to pass.
> 
> This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1930)

Is there a typo in the issue number? This does not seem to be very relevant to #898

File: sdk/python/kfp/dsl/_pipeline_param.py (hunk 1 above)
Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline_param.py]: Good idea. Added the type.

**Comment by Ark-kun** [comment]
> Is there a typo in the issue number?

Good catch. My copy-paste buffer malfunctioned. It should be https://github.com/kubeflow/pipelines/issues/1420

Fixed.

**Comment by gaoning777** [comment]
Why are we deleting the TypeMeta? The internal type representation is useful as a strongly typed one.

**Comment by Ark-kun** [comment]
> Why are we deleting the TypeMeta? The internal type representation is useful as a strongly typed one.

It's part of the unification of the almost duplicate classes `ComponentSpec`, `ComponentMeta`, `PipelineMeta`.
https://github.com/kubeflow/pipelines/blob/7b6825d7d50913d0a5c91df032d8cfba1ec7c324/sdk/python/kfp/dsl/_metadata.py#L74

`TypeMeta` was a bit different from `ParameterMeta`, `ComponentMeta` or `PipelineMeta`, because it's serialization was different. Without `TypeMeta` the unification becomes almost as trivial as renaming. Also `TypeMeta` was an internal class (as opposed to the types in `kfp.dsl.types`), so all the ways it's created and consumed were controlled by the SDK internals (the type dictionary structure is checked in several places).

The internal representation of different component parts are very strongly typed and are fully type-checked during the construction, so the end result will be as strongly-typed as before is not more strongly-typed.

In the end, the unification will help us unify shareable components, ContainerOps and pipelines.
Right now there are three different code bases for them that are very similar.
For example the default value type checking works different between `@pipeline`, `@component` and `load_component`.

**Comment by gaoning777** [comment]
Discussed offline, will add the TypeMeta back to the component module. 
Another effort will be made to converge the other *meta classes such as ParameterMeta, etc.

**Pull request merged**